### PR TITLE
Make Yul optimizer not fail for wasm.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libyul/CompilabilityChecker.cpp [hunk 1]
```diff
@@ -43,27 +43,31 @@ map<YulString, int> CompilabilityChecker::run(
...
-	CodeTransform transform(assembly, analysisInfo, _ast, noOutputDialect, builtinContext, _optimizeStackAllocation);
-	try
-	{
-		transform(_ast);
-	}
-	catch (StackTooDeepError const&)
-	{
-		solAssert(!transform.stackErrors().empty(), "Got stack too deep exception that was not stored.");
-	}
+		NoOutputAssembly assembly;
```
Comment (bshastry) [inline libyul/CompilabilityChecker.cpp:54]: whitespace?

File: libyul/CompilabilityChecker.cpp [hunk 2]
```diff
@@ -43,27 +43,31 @@ map<YulString, int> CompilabilityChecker::run(
...
-	NoOutputEVMDialect noOutputDialect(dynamic_cast<EVMDialect const&>(_dialect));
-	BuiltinContext builtinContext;
+	if (EVMDialect const* evmDialect = dynamic_cast<EVMDialect const*>(&_dialect))
+	{
+		NoOutputEVMDialect noOutputDialect(*evmDialect);
+		BuiltinContext builtinContext;
 
-	yul::AsmAnalysisInfo analysisInfo =
-		yul::AsmAnalyzer::analyzeStrictAssertCorrect(noOutputDialect, _ast);
+		yul::AsmAnalysisInfo analysisInfo =
```
Comment (bshastry) [inline libyul/CompilabilityChecker.cpp:51]: whitespace here and below.

File: libyul/AssemblyStack.cpp [hunk 3]
```diff
@@ -146,15 +147,17 @@ void AssemblyStack::optimize(Object& _object, bool _isCreation)
...
+	{
+		GasMeter meter(*dialect, _isCreation, m_optimiserSettings.expectedExecutionsPerDeployment);
+		OptimiserSuite::run(
+			*dialect,
+			meter,
+			*_object.code,
+			*_object.analysisInfo,
+			m_optimiserSettings.optimizeStackAllocation
+		);
+	}
```
Comment (Marenz) [inline libyul/AssemblyStack.cpp:160]: The intention here is to not run the optimizer if it's set to a non-evm dialect?

**Comment by leonardoalt** [comment]
Is `optimiser/Metrics.*` still being used anywhere?

File: libyul/CompilabilityChecker.cpp (hunk 2 above)
Comment (leonardoalt) [reply libyul/CompilabilityChecker.cpp:51]: That's because the assignment line would be too long

File: libyul/AssemblyStack.cpp (hunk 3 above)
Comment (chriseth) [reply libyul/AssemblyStack.cpp:160]: Yes, not run the constant optimizer, because it only works on EVM for now.

**Comment by chriseth** [comment]
@leonardoalt yes, it still contains the non-evm-specific metrics.

**Comment by leonardoalt** [comment]
It looks good to me, but emscripten build is failing

**Comment by bshastry** [comment]
Emscripten build fails due to linking errors. Linker cannot locate the following function signatures

```
yul::GasMeter::costs(boost::variant<yul::FunctionalInstruction, yul::FunctionCall, yul::Identifier, yul::Literal> const&) const
yul::GasMeter::instructionCosts(dev::eth::Instruction) const
```

**Comment by chriseth** [comment]
It seems the compiler cannot handle files with the same name but in different directories.

**Review by leonardoalt — DISMISSED** [review]

File: libyul/optimiser/ControlFlowSimplifier.cpp [hunk 4]
```diff
@@ -33,11 +34,15 @@ using OptionalStatements = boost::optional<vector<Statement>>;
 namespace
 {
 
-ExpressionStatement makePopExpressionStatement(langutil::SourceLocation const& _location, Expression&& _expression)
+ExpressionStatement makePopExpressionStatement(
+	langutil::SourceLocation const& _location,
+	YulString _popName,
```
Comment (ekpyron) [inline libyul/optimiser/ControlFlowSimplifier.cpp]: I'd say we should either pass in the dialect and get the ``discardFunction`` from it inside this function (it's only called where the dialect is available) or we should rename it to something like ``makeUnaryFunctionCallExpressionStatement``... but it's not too important...

We could also make this and all the other free functions here that now require a dialect argument private members of ``ControlFlowSimplifier`` and use ``m_dialect`` instead - or put them in an implementation-local ``Util`` class with a dialect member... not sure either is worth it, though.

File: libyul/backends/wasm/WasmDialect.h [hunk 5]
```diff
@@ -45,6 +45,8 @@ struct WasmDialect: public Dialect
 	WasmDialect();
 
 	BuiltinFunction const* builtin(YulString _name) const override;
+	BuiltinFunction const* discardFunction() const override { return builtin("drop"_yulstring); }
+	BuiltinFunction const* equalityFunction() const override { return builtin("i64.eq"_yulstring); }
```
Comment (ekpyron) [inline libyul/backends/wasm/WasmDialect.h:49]: How can we use a 64-bit comparison here where we have a 256-bit comparison for EVM? It's used for ``switch``->``if`` below - will WASM ``switch``es be 64-bit and not simulate 256 bits?

File: libyul/optimiser/ControlFlowSimplifier.cpp (hunk 4 above)
Comment (chriseth) [reply libyul/optimiser/ControlFlowSimplifier.cpp]: I'll change it to use the dialect.

File: libyul/backends/wasm/WasmDialect.h (hunk 5 above)
Comment (chriseth) [reply libyul/backends/wasm/WasmDialect.h:49]: This dialect only knows 64 bit types (as per the comment on the class).

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**